### ISSUE TITLE
expect window.GM_api but got undefined if code not include .+\.GM

### ISSUE THREAD

**Issue by lisonge** [Tampermonkey/tampermonkey#1567]
hello, I am developing [vite-plugin-monkey](https://github.com/lisonge/vite-plugin-monkey) to quickly develop UserScript

I want to inject GM_api to document then code can use api by specific ESM exports

but when I get window.GM_api, sometimes I will get undefined, the following is example

## Script

(Please give an example of the script if applicable.)

```js
// ==UserScript==
// @name               dev:example
// @namespace          https://github.com/lisonge
// @version            1.0.1
// @match              https://i.songe.li/
// @grant              GM.addElement
// @grant              GM_openInTab
// @grant              GM_cookie
// ==/UserScript==

;(() => {
console.log(Reflect.get(window, 'GM')) // ouput undefined
   console.log(
  Object.entries(window)
    .filter(([k]) => k.startsWith('GM'))
    .map(([k]) => k),
); // output -> []
})();
```
an example of the script if applicable.)

```js
// ==UserScript==
// @name               dev:example
// @namespace          https://github.com/lisonge
// @version            1.0.1
// @match              https://i.songe.li/
// @grant              GM.addElement
// @grant              GM_openInTab
// @grant              GM_cookie
// ==/UserScript==

;(() => {
// if the comment exists
  //any_anme.GM
console.log(Reflect.get(window, 'GM')) // ouput Object
   console.log(
  Object.entries(window)
    .filter(([k]) => k.startsWith('GM'))
    .map(([k]) => k),
); // output ->  ['GM', 'GM_openInTab', 'GM_cookie']
})();
```

when code does not include `.+\.GM` ,  why I can not get window.GM_api ?

**Comment by derjanb** [comment]
You can configure this behavior using the "Add GM functions to this or window" setting for each script.

**Comment by lisonge** [comment]
@derjanb 

Maybe it is indeed a bug

error example

```js
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @match        https://*
// @grant        GM_getValue
// ==/UserScript==

(function () {
    'use strict';
    var Q = window,
        mt = (() => Q.GM_getValue)();
    console.log([Q,mt]);
})();
```
![image](https://user-images.githubusercontent.com/38517192/224490268-6572e22a-65d3-45af-ba03-95d930f2dea5.png)

---
work axample，`just add a comment`
```js
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @match        https://*
// @grant        GM_getValue
// ==/UserScript==

(function () {
    'use strict';
    var Q = window,
        mt = (() => Q.GM_getValue)();
    console.log([Q,mt]);
    // window.GM_getValue
})();

```
![image](https://user-images.githubusercontent.com/38517192/224490328-34effeb7-e06c-4231-b52a-9506131b1b80.png)

**Comment by lisonge** [comment]
This bug comes from [vite-plugin-monkey#55](https://github.com/lisonge/vite-plugin-monkey/discussions/55)

**Comment by lisonge** [comment]
emmm, If it is not a bug, what is the significance of this behavior?

**Comment by Malix-Off** [comment]
Note **this issue causes minificated build using vite be unusable by tampermonkey** ([exemple](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5275639)).

**Comment by Malix-Off** [comment]
Further bug reproduction exemple from @lisonge: https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5276119

**Comment by derjanb** [comment]
> > You can configure this behavior using the `Add GM functions to this or window` setting for each script.
> 
> emmm, If it is not a bug, what is the significance of this behavior?

Userscripts sometime pass unsafeWindow to `Function.prototype.apply` which can be overridden by the page. That's why Tampermonkey attaches these powerful functions only if `<something>.GM_getValue` is detected. Otherwise it is only accessible via `GM_getValue`. That's why the comment helps... Also see the [changelog](https://www.tampermonkey.net/changelog.php?more=true&show=gcal#v4.16.6158).

`<something>` was expected to be `window` for example, but the detection is more general. However it doesn't match the example above. I'll fix that.

As I said above: the automatic detection can be disabled via `Add GM functions to this or window` at the script's settings.

**Comment by Malix-Off** [comment]
@derjanb, note that @lisonge is the author of [`vite-plugin-monkey`](https://www.npmjs.com/package/vite-plugin-monkey).

He's basically trying to make his plugin output a build file that would actually be working with Tampermonkey, that is not currently the case.

**Comment by derjanb** [comment]
Should be fixed at 4.19.6180 ([crx](https://tampermonkey.net/crx/tampermonkey_beta_current.crx)|[xpi](https://firefox.tampermonkey.net/firefox-current-beta.xpi))

Chrome/Edge users, please download the crx file linked above and drag and drop it to the extensions page `chrome://extensions`.
Firefox users please install the [BETA version](https://www.tampermonkey.net/index.php?browser=firefox) or check for BETA version updates at `about:addons` 

For a quick fix please export your settings and scripts as zip or (JSON) file at the "Utilities" tab and import it back at the fixed BETA version.

**Comment by Malix-Off** [comment]
@derjanb:
> Should be fixed at 4.19.6180 ([crx](https://tampermonkey.net/crx/tampermonkey_beta_current.crx)|[xpi](https://firefox.tampermonkey.net/firefox-current-beta.xpi))
> 
> Chrome/Edge users, please download the crx file linked above and drag and drop it to the extensions page `chrome://extensions`. Firefox users please install the [BETA version](https://www.tampermonkey.net/index.php?browser=firefox) or check for BETA version updates at `about:addons`
> 
> For a quick fix please export your settings and scripts as zip or (JSON) file at the "Utilities" tab and import it back at the fixed BETA version.

Do you've got any ETA on the gold release?
